pkg/xds/generator/outbound_proxy_generator.go

```diff
@@ -27,142 +29,172 @@ func (g OutboundProxyGenerator) Generate(ctx xds_context.Context, proxy *model.P
+				Configure(envoy_listeners.NetworkAccessLog(meshName, envoy_listeners.TrafficDirectionOutbound, sourceService, serviceName, proxy.Logs[serviceName], proxy))
 		}
-		clusters = append(clusters, envoy_common.ClusterInfo{
-			Name:   envoy_names.GetDestinationClusterName(service, destination.Destination),
-			Weight: destination.Weight,
-			Tags:   destination.Destination,
-		})
+		return filterChainBuilder
+	}()
+	listener, err := envoy_listeners.NewListenerBuilder().
+		Configure(envoy_listeners.OutboundListener(outboundListenerName, oface.DataplaneIP, oface.DataplanePort)).
+		Configure(envoy_listeners.FilterChain(filterChainBuilder)).
+		Configure(envoy_listeners.TransparentProxying(proxy.Dataplane.Spec.Networking.GetTransparentProxying())).
+		Build()
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not generate listener %s for service %s", outboundListenerName, serviceName)
 	}
-	return
+	return listener, nil
 }
 
-func (_ OutboundProxyGenerator) generateEds(ctx xds_context.Context, proxy *model.Proxy, clusters []envoy_common.ClusterInfo) (resources []*model.Resource, allEndpoints []model.Endpoint, _ error) {
-	for _, cluster := range clusters {
-		serviceName := cluster.Tags[kuma_mesh.ServiceTag]
+func (_ OutboundProxyGenerator) generateCDS(ctx xds_context.Context, proxy *model.Proxy, clusters envoy_common.Clusters) (model.ResourceSet, error) {
+	var resources model.ResourceSet
+	for _, clusterName := range clusters.ClusterNames() {
+		serviceName := clusters.Tags(clusterName)[0][kuma_mesh.ServiceTag]
 		healthCheck := proxy.HealthChecks[serviceName]
 		circuitBreaker := proxy.CircuitBreakers[serviceName]
 		edsCluster, err := envoy_clusters.NewClusterBuilder().
-			Configure(envoy_clusters.EdsCluster(cluster.Name)).
-			Configure(envoy_clusters.ClientSideMTLS(ctx, proxy.Metadata, cluster.Tags[kuma_mesh.ServiceTag])).
+			Configure(envoy_clusters.EdsCluster(clusterName)).
+			Configure(envoy_clusters.LbSubset(clusters.Tags(clusterName))).
+			Configure(envoy_clusters.ClientSideMTLS(ctx, proxy.Metadata, serviceName)).
 			Configure(envoy_clusters.OutlierDetection(circuitBreaker)).
 			Configure(envoy_clusters.HealthCheck(healthCheck)).
 			Build()
 		if err != nil {
-			return nil, nil, err
+			return model.ResourceSet{}, err
 		}
-		resources = append(resources, &model.Resource{
-			Name:     cluster.Name,
-			Resource: edsCluster,
-		})
-		endpoints := model.EndpointList(proxy.OutboundTargets[serviceName]).Filter(cluster.Tags)
-		resources = append(resources, &model.Resource{
-			Name:     cluster.Name,
-			Resource: envoy_endpoints.CreateClusterLoadAssignment(cluster.Name, endpoints),
+		resources.AddNamed(edsCluster)
+	}
+	return resources, nil
+}
+
+func (_ OutboundProxyGenerator) generateEDS(proxy *model.Proxy, clusters envoy_common.Clusters) model.ResourceSet {
+	var resources model.ResourceSet
+	for _, clusterName := range clusters.ClusterNames() {
+		serviceName := clusters.Tags(clusterName)[0][kuma_mesh.ServiceTag]
```

Comment:
Does `serviceName` equal to `clusterName`?

Reply:
For now it does, but just for safety I don't want to make this assumption.